group-coordinator/src/main/java/org/apache/kafka/coordinator/group/util/SystemTimerReaper.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.util;
+
+import org.apache.kafka.server.util.ShutdownableThread;
+import org.apache.kafka.server.util.timer.Timer;
+import org.apache.kafka.server.util.timer.TimerTask;
+
+/**
+ * SystemTimerReaper wraps a {@link Timer} and starts a reaper thread
+ * to expire the tasks in the {@link Timer}.
+ */
+public class SystemTimerReaper implements Timer {
+    private static final long WORK_TIMEOUT_MS = 200L;
```

Comment:
to confirm, we don't actually advance the timer by 200ms every iteration but wait for a max 200ms if there are no tasks to expire.

this is more of a question on system timer, but in `SystemTimer#advanceClock()`
```
timingWheel.advanceClock(bucket.getExpiration());
```
how is the timer in-sync with the actual time if we advance clock by a custom amount?

Reply:
Yeah. I got confused by this as well. That was my understanding. That the clock doesn't actually advance unless we expire an event. And we only advance it to that time.

Reply:
My understanding is that if the clock is advanced by a custom amount, the items in the `DelayedQueue` will be expired and the next call to `advanceClock` will process them.

Reply:
so does the delay queue store buckets that are ready to be expired? if that's the case it makes sense.

Reply:
i suppose so but i don't know the timing wheel very well.

Reply:
I'm a bit uncertain about this change, and the uncertainty from this thread doesn't inspire confidence. 😅

Reply:
The important part is that it works like the other reapers that I linked in another comment.

Reply:
https://github.com/apache/kafka/blob/trunk/server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java has a good javadoc if you want to read more about it.

Reply:
I read some of this, but I guess I'm just not convinced you understood it when this code was written. Hence my concern on the code. I guess I will read it again.

Reply:
Haha. Are you saying that I don’t understand what I am doing?

Reply:
I understand that the system timer and the timing wheel are not easy to grasp when you don't know them. I also realize that my reply to Jeff [here](https://github.com/apache/kafka/pull/13991#discussion_r1262824134) was poorly written. I should have been sharper on this one in order to reduce the confusion.

Let me try to re-explain how this works in order to reduce the confusion here. As Jeff said, the time is not necessarily advanced by 200ms every time the thread ticks. The 200ms is used to wait on the delayed queue which contains the buckets managed by the timing wheel. This means that it waits maximum 200ms for a bucket to expire or return. When a bucket expires, the time of the timing wheel is advanced to the expiration time of the bucket and all the timer tasks in that bucket are handled. Note that handled here does not mean expired because the bucket could be from the "overflow wheel". In this case, they may have to be added back to the current timing wheel or expired.

Reply:
Thanks. I didn't mean to imply that so I'm sorry it came across that way.